### DISCUSSION TITLE
Event subscription

### DISCUSSION THREAD

**Discussion by FShinoda** [FreeOpcUa/opcua-asyncio#1274] [Q&A] [answered]
Hello there!

I'm currently in a project that uses opc-ua protocol and I've been trying to use this library to subscribe to events and watch for node data changes.

I'm facing some difficulties handling the events incoming from the subscription. 
I am trying to print the current datetime using `datetime.now()` function and then print the incoming event.
``` python
class SubHandler:
    def event_notification(self, event):
          print(datetime.datetime.now())
          print("Python: New event", event)
```
But, what actually outputs is this:

![image](https://user-images.githubusercontent.com/45886482/230728574-79968575-a126-4ced-83e1-aeabf87f2202.png)

So, it seems like it never enters the `event_notification` function...

Am I missing something?

Here's my full code:

``` python
import asyncio
import logging
import time
import datetime
from asyncua import Client, Node, ua
from asyncua.crypto.security_policies import SecurityPolicyBasic256Sha256
from asyncua.common.events import Event
logging.basicConfig(level=logging.INFO)
_logger = logging.getLogger(__name__)


class SubHandler:
    """
    The SubscriptionHandler is used to handle the data that is received for the subscription.
    """
    def datachange_notification(self, node: Node, val, data):
        """
        Callback for asyncua Subscription.
        This method will be called when the Client received a data change message from the Server.
        """
        
        print("Python: New data change event", node, val)
        
    def event_notification(self, event):
        
        print(datetime.datetime.now())
        print("Python: New data change event", event)
        

async def task():
    ## Set project URL
    url = "opc.tcp://my_ip_here"
    client = Client(url=url, timeout=10)
    
    client.set_user('my_user_here')
    client.set_password('my_password_here')
    
    await client.set_security(
        SecurityPolicyBasic256Sha256,
        certificate='./certificados/google_certificado.pem',
        private_key='./certificados/google_chave_privada.pem',
        server_certificate="./certificados/sage_certificado.der"
    )

    async with client: 
        await client.open_secure_channel(renew=False)
    
        print("#"*25)
        ## Get root node
        root_node = client.get_root_node()
        print("The root node id is:", root_node)
        
        # Subscribe to events on the node
        handler = SubHandler()
        subscription = await client.create_subscription(500, handler)
        await subscription.subscribe_events()
                    
    
if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    asyncio.run(task())
```

**Comment by schroeder-** [comment]
Which version are you using? We just released a version to pip, that fixed some issues with subscript to events.

**Comment by FShinoda** [comment]
I upgraded the package using `pip install asyncua --upgrade` and it installed the version `1.0.2`. 

![image](https://user-images.githubusercontent.com/45886482/230735695-63950f6c-7ee7-4629-ab59-0a3c89b54418.png)

Still, I am getting the same output:
![image](https://user-images.githubusercontent.com/45886482/230735705-72fc54ad-faf1-401a-997d-f9c40a97a3c6.png)

**Reply by schroeder-** [reply]
Did you check via Ua Expert that the server send Events?

**Reply by FShinoda** [reply]
This is something I don't have acess. I will ask someone else and, then, I return you the answer. 
Thank you, for your help so far.

**Reply by GoetzGoerisch** [reply]
This sample server sends events:
https://github.com/umati/Sample-Server

Running instance at opc.tcp://opcua.umati.app:4840

**Reply by GoetzGoerisch** [reply]
UaExpert to be downloaded here: https://www.unified-automation.com/products/development-tools/uaexpert.html
Free registration required.

**Reply by FShinoda** [reply]
> Did you check via Ua Expert that the server send Events?

I've just talked to my client about that. They gave me this other node to try it out:

<img width="784" alt="image" src="https://user-images.githubusercontent.com/45886482/230971765-be3f00b7-4732-41db-910e-13359d9ae42b.png">
<img width="319" alt="image" src="https://user-images.githubusercontent.com/45886482/230973962-2df319d9-1a54-4516-85e9-0595d59475ca.png">

 They claim this node changes its value frequently, so I think I should be using the `subscribe_data_change` subscription, right?
 But, using that, I am currently facing this issue:

> ERROR:asyncua.crypto.uacrypto:certificate does not cotain the application uri (urn:freeopcua:client). Most applications will reject a connection without it.
> ERROR:asyncua.crypto.uacrypto:certificate does not cotain the hostname in DNSNames python-20230328-140326. Some applications will check this.
> INFO:asyncua.client.client:connect
> INFO:asyncua.client.ua_client.UaClient:opening connection
> INFO:asyncua.uaprotocol:updating client limits to: TransportLimits(max_recv_buffer=65536, max_send_buffer=65536, max_chunk_count=0, max_message_size=0)
> INFO:asyncua.client.ua_client.UASocketProtocol:open_secure_channel
> INFO:asyncua.client.client:Requested secure channel timeout to be 3600000ms, got 1800000ms instead
> INFO:asyncua.client.ua_client.UaClient:create_session
> INFO:asyncua.client.ua_client.UaClient:activate_session
> INFO:asyncua.client.ua_client.UASocketProtocol:open_secure_channel
> #########################
> The root node id is: i=84
> INFO:asyncua.client.ua_client.UaClient:create_subscription success SubscriptionId 1
> INFO:asyncua.common.subscription:Subscription created 1
> WARNING:asyncua.client.client:Revised values returned differ from subscription values: CreateSubscriptionResult(SubscriptionId=1, RevisedPublishingInterval=500.0, RevisedLifetimeCount=3600, RevisedMaxKeepAliveCount=10)
> INFO:asyncua.client.ua_client.UaClient:create_monitored_items
> INFO:asyncua.client.client:disconnect
> ERROR:asyncua.client.client:Error in watchdog loop
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.7/site-packages/asyncua/client/client.py", line 518, in _monitor_server_loop
>     await asyncio.sleep(timeout)
>   File "/opt/conda/lib/python3.7/asyncio/tasks.py", line 595, in sleep
>     return await future
> concurrent.futures._base.CancelledError
> INFO:asyncua.common.subscription:Publish callback called with result: PublishResult(SubscriptionId=1, AvailableSequenceNumbers=[], MoreNotifications=True, NotificationMessage_=NotificationMessage(SequenceNumber=0, PublishTime=datetime.datetime(2023, 4, 10, 19, 11, 29, 672651), NotificationData=[StatusChangeNotification(Status=2148270080, DiagnosticInfo_=DiagnosticInfo(SymbolicId=None, NamespaceURI=None, Locale=None, LocalizedText=None, AdditionalInfo=None, InnerStatusCode=None, InnerDiagnosticInfo=None))]), Results=[], DiagnosticInfos=[])
> STATUS: StatusChangeNotification(Status=2148270080, DiagnosticInfo_=DiagnosticInfo(SymbolicId=None, NamespaceURI=None, Locale=None, LocalizedText=None, AdditionalInfo=None, InnerStatusCode=None, InnerDiagnosticInfo=None))
> ERROR:asyncua.client.client:Error while renewing session
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.7/site-packages/asyncua/client/client.py", line 541, in _renew_channel_loop
>     await asyncio.sleep(duration)
>   File "/opt/conda/lib/python3.7/asyncio/tasks.py", line 595, in sleep
>     return await future
> concurrent.futures._base.CancelledError
> INFO:asyncua.client.ua_client.UaClient:close_session
> INFO:asyncua.client.ua_client.UASocketProtocol:close_secure_channel
> INFO:asyncua.client.ua_client.UASocketProtocol:Request to close socket received
> INFO:asyncua.client.ua_client.UASocketProtocol:Socket has closed connection

and when I use `subscribe_events` , this one:

> ERROR:asyncua.crypto.uacrypto:certificate does not cotain the application uri (urn:freeopcua:client). Most applications will reject a connection without it.
> ERROR:asyncua.crypto.uacrypto:certificate does not cotain the hostname in DNSNames python-20230328-140326. Some applications will check this.
> INFO:asyncua.client.client:connect
> INFO:asyncua.client.ua_client.UaClient:opening connection
> INFO:asyncua.uaprotocol:updating client limits to: TransportLimits(max_recv_buffer=65536, max_send_buffer=65536, max_chunk_count=0, max_message_size=0)
> INFO:asyncua.client.ua_client.UASocketProtocol:open_secure_channel
> INFO:asyncua.client.client:Requested secure channel timeout to be 3600000ms, got 1800000ms instead
> INFO:asyncua.client.ua_client.UaClient:create_session
> INFO:asyncua.client.ua_client.UaClient:activate_session
> INFO:asyncua.client.ua_client.UASocketProtocol:open_secure_channel
> #########################
> The root node id is: i=84
> INFO:asyncua.client.ua_client.UaClient:create_subscription success SubscriptionId 1
> INFO:asyncua.common.subscription:Subscription created 1
> WARNING:asyncua.client.client:Revised values returned differ from subscription values: CreateSubscriptionResult(SubscriptionId=1, RevisedPublishingInterval=500.0, RevisedLifetimeCount=3600, RevisedMaxKeepAliveCount=10)
> INFO:asyncua.client.ua_client.UaClient:browse
> INFO:asyncua.client.ua_client.UaClient:create_monitored_items
> INFO:asyncua.common.subscription:Publish callback called with result: PublishResult(SubscriptionId=1, AvailableSequenceNumbers=[], MoreNotifications=False, NotificationMessage_=NotificationMessage(SequenceNumber=0, PublishTime=datetime.datetime(2023, 4, 10, 19, 9, 1, 49329), NotificationData=[]), Results=[], DiagnosticInfos=[])
> INFO:asyncua.client.client:disconnect
> ERROR:asyncua.client.client:Error in watchdog loop
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.7/site-packages/asyncua/client/client.py", line 518, in _monitor_server_loop
>     await asyncio.sleep(timeout)
>   File "/opt/conda/lib/python3.7/asyncio/tasks.py", line 595, in sleep
>     return await future
> concurrent.futures._base.CancelledError
> INFO:asyncua.common.subscription:Publish callback called with result: PublishResult(SubscriptionId=1, AvailableSequenceNumbers=[], MoreNotifications=True, NotificationMessage_=NotificationMessage(SequenceNumber=0, PublishTime=datetime.datetime(2023, 4, 10, 19, 9, 1, 283706), NotificationData=[StatusChangeNotification(Status=2148270080, DiagnosticInfo_=DiagnosticInfo(SymbolicId=None, NamespaceURI=None, Locale=None, LocalizedText=None, AdditionalInfo=None, InnerStatusCode=None, InnerDiagnosticInfo=None))]), Results=[], DiagnosticInfos=[])
> STATUS: StatusChangeNotification(Status=2148270080, DiagnosticInfo_=DiagnosticInfo(SymbolicId=None, NamespaceURI=None, Locale=None, LocalizedText=None, AdditionalInfo=None, InnerStatusCode=None, InnerDiagnosticInfo=None))
> ERROR:asyncua.client.client:Error while renewing session
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.7/site-packages/asyncua/client/client.py", line 541, in _renew_channel_loop
>     await asyncio.sleep(duration)
>   File "/opt/conda/lib/python3.7/asyncio/tasks.py", line 595, in sleep
>     return await future
> concurrent.futures._base.CancelledError
> INFO:asyncua.client.ua_client.UaClient:close_session
> INFO:asyncua.client.ua_client.UASocketProtocol:close_secure_channel
> INFO:asyncua.client.ua_client.UASocketProtocol:Request to close socket received
> INFO:asyncua.client.ua_client.UASocketProtocol:Socket has closed connection
> Traceback (most recent call last):
>   File "teste_subscribe.py", line 68, in <module>
>     asyncio.run(task())
>   File "/opt/conda/lib/python3.7/asyncio/runners.py", line 43, in run
>     return loop.run_until_complete(main)
>   File "/opt/conda/lib/python3.7/asyncio/base_events.py", line 587, in run_until_complete
>     return future.result()
>   File "teste_subscribe.py", line 63, in task
>     await subscription.subscribe_events(exam)
>   File "/opt/conda/lib/python3.7/site-packages/asyncua/common/subscription.py", line 249, in subscribe_events
>     return await self._subscribe(sourcenode, ua.AttributeIds.EventNotifier, evfilter, queuesize=queuesize)  # type: ignore
>   File "/opt/conda/lib/python3.7/site-packages/asyncua/common/subscription.py", line 323, in _subscribe
>     mids[0].check()
>   File "/opt/conda/lib/python3.7/site-packages/asyncua/ua/uatypes.py", line 372, in check
>     raise UaStatusCodeError(self.value)
> **asyncua.ua.uaerrors._auto.BadNotReadable: "The access level does not allow reading or subscribing to the Node."(BadNotReadable)**
> 

Totally different outputs by just changing one line, therefore I don't think this is a connection problem... 
Can you help me?

My code:
```python
import asyncio
import logging
import time
import datetime
from asyncua import Client, Node, ua
from asyncua.ua.uatypes import NodeId
from asyncua.crypto.security_policies import SecurityPolicyBasic256Sha256, MessageSecurityMode
from asyncua.common.events import Event
logging.basicConfig(level=logging.INFO)
_logger = logging.getLogger(__name__)


class SubHandler:
    """
    The SubscriptionHandler is used to handle the data that is received for the subscription.
    """
    def datachange_notification(self, node: Node, val, data):
        """
        Callback for asyncua Subscription.
        This method will be called when the Client received a data change message from the Server.
        """
        print(datetime.datetime.now())
        print("Python: New data change event", node, val)
        
    def event_notification(self, event):
        
        print(datetime.datetime.now())
        print("Python: New data change event", event)
        
    def status_change_notification(self, status):
        print('STATUS:', status)
        

async def task():
    ## Set project URL
    url = "opc.tcp://my_ip:4840"
    client = Client(url=url, timeout=10)
    
    client.set_user('my_user')
    client.set_password('my_password')
    
    await client.set_security(
        SecurityPolicyBasic256Sha256,
        mode= MessageSecurityMode.SignAndEncrypt,
        certificate='./certificados/google_certificado.pem',
        private_key='./certificados/google_chave_privada.pem',
        server_certificate="./certificados/sage_certificado.der"
    )

    async with client: 
        await client.open_secure_channel()
    
        print("#"*25)
        ## Get root node
        root_node = client.get_root_node()
        print("The root node id is:", root_node)
        
        # Subscribe to events on the node
        exam = client.get_node('ns=1; s=BLBSLI403MIFA.pas.valor')
        handler = SubHandler()
        subscription = await client.create_subscription(500, handler)
        # await subscription.subscribe_data_change(exam)
        await subscription.subscribe_events(exam)
                    
    
if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    asyncio.run(task())
```

**Reply by schroeder-** [reply, marked as answer]
If you want to get a data change then call `subscribe_data_change`, events in opc ua is something totally different.
You directly disconnect in your code. This code is what you want:
``` Python
 async with client: 
        print("#"*25)
        ## Get root node
        root_node = client.get_root_node()
        print("The root node id is:", root_node)
        
        # Subscribe to events on the node
        exam = client.get_node('ns=1; s=BLBSLI403MIFA.pas.valor')
        handler = SubHandler()
        subscription = await client.create_subscription(500, handler)
        await subscription.subscribe_data_change(exam)
        # await subscription.subscribe_events(exam)
        while True:
               await asyncio.sleep(1.0)
```

**Reply by FShinoda** [reply]
Thanks! It finally works 😊

**Comment by xyzakdkd** [comment]
how to close the subscription depending on a value of a node in the code above

**Reply by schroeder-** [reply]
Add the subscription to your SubHandler. IIn datachange_notification you can unsubscripe the value with a call to subscription.unsubscribe(data.subscription_data.client_handle)

```
class SubHandler:
    subscription = None
    """
    The SubscriptionHandler is used to handle the data that is received for the subscription.
    """
    async def datachange_notification(self, node: Node, val, data):
        """
        Callback for asyncua Subscription.
        This method will be called when the Client received a data change message from the Server.
        """
        if val == ....:
             await subscription.unsubscribe(data.subscription_data.client_handle)

    ....
    async with client: 
        # Subscribe to events on the node
        exam = client.get_node('ns=1; s=BLBSLI403MIFA.pas.valor')
        handler = SubHandler()
        subscription = await client.create_subscription(500, handler)
        # Init the subscription in the handler
        handler.subscription = subscription
        await subscription.subscribe_data_change(exam)
        while True:
               await asyncio.sleep(1.0)
```

**Reply by xyzakdkd** [reply]
thanks for the fast answer. i  try it but it show me an Error : Exception calling data change handler. Error: 'NoneType' object has no attribute 'unsubscribe'

**my main** 

    handler = subhandler.SubHandler()
    subscription = await client.create_subscription(500, handler)
    handler.subscription = subscription
    subscriptions = []


 **my subHandler class**

```
class SubHandler:
    subscription = None
    def __init__(self):
        self.start_up = datetime.utcnow()
        self.souce_timestamp_start =False

    async def datachange_notification(self, node: Node, val, data):
        node = str(node)
        if (self.CurrentResult == "Successful") or (self.CurrentResult == "UserCancelled") or (self.CurrentResult == None):
            print("The process has been SUCCESSFUL finished or it has been CANCELLED by the USER")
            await SubHandler.subscription.unsubscribe(data.subscription_data.client_handle)
```

every thing work with me good but only the park where i unsubscibe.

**Reply by schroeder-** [reply]
`await self.subscription.unsubscribe(data.subscription_data.client_handle)`

**Reply by xyzakdkd** [reply]
i try it and it didnt work cound the Error version iam using asyncua 1.0.4


`The process has been SUCCESSFUL finished or it has been CANCELLED by the USER
Exception calling data change handler. Error: The monitoring item id does not refer to a valid monitored item.(BadMonitoredItemIdInvalid)
Traceback (most recent call last):
  File "C:\Users\love\AppData\Local\Programs\Python\Python39\lib\site-packages\asyncua\common\subscription.py", line 152, in _call_datachange
    await asyncio.gather(*tasks)
  File "c:\Users\love\Pictures\5-asyncua_subscription(all 4 topics)\app\subhandler.py", line 81, in datachange_notification
    await self.subscription.unsubscribe(data.subscription_data.client_handle)
  File "C:\Users\love\AppData\Local\Programs\Python\Python39\lib\site-packages\asyncua\common\subscription.py", line 394, in unsubscribe
    results[0].check()
  File "C:\Users\love\AppData\Local\Programs\Python\Python39\lib\site-packages\asyncua\ua\uatypes.py", line 373, in check
    raise UaStatusCodeError(self.value)
asyncua.ua.uaerrors._auto.BadMonitoredItemIdInvalid: The monitoring item id does not refer to a valid monitored item.(BadMonitoredItemIdInvalid)
currentResult None`

**Reply by schroeder-** [reply]
This should be the correct value
`await self.subscription.unsubscribe(data.subscription_data.server_handle)`

**Reply by xyzakdkd** [reply]
thanks it work
